tarjonta-service/src/main/java/fi/vm/sade/tarjonta/dao/impl/HakuDAOImpl.java

```diff
@@ -357,4 +357,30 @@ public void safeDelete(final String hakuOid, final String userOid) {
           .where(qHaku.tila.ne(TarjontaTila.POISTETTU))
           .list(qHaku);
     }
+
+    @Override
+    public List<Haku> findByAlkamisvuosi(Integer alkamisVuosi) {
+
+        String q = ("select distinct(h3.*) from koulutusmoduuli_toteutus komoto " +
+                "join koulutus_hakukohde kh on komoto.id = kh.koulutus_id " +
+                "join hakukohde h2 on kh.hakukohde_id = h2.id " +
+                "join haku h3 on h2.haku_id = h3.id " +
+                "where komoto.alkamisvuosi is not null " +
+                "and h2.tila = 'JULKAISTU' " +
+                "and komoto.tila = 'JULKAISTU' " +
+                "and h3.koulutuksen_alkamisvuosi is null " +
+                "and komoto.alkamisvuosi = ? " +
+                "union " +
+                "select distinct(h3.*) from koulutusmoduuli_toteutus komoto " +
+                "join koulutus_hakukohde kh on komoto.id = kh.koulutus_id " +
+                "join hakukohde h2 on kh.hakukohde_id = h2.id " +
+                "join haku h3 on h2.haku_id = h3.id " +
+                "where komoto.tila = 'JULKAISTU' " +
+                "and h2.tila = 'JULKAISTU' " +
+                "and h3.koulutuksen_alkamisvuosi = ?");
```

Comment:
Olisi ehkä helpompi lukea jos lyhenteenä olisi `hk` ja `h` eikä `h2` ja `h3`.

Tämä: `"where komoto.alkamisvuosi is not null "` lienee jo sisältyneenä tähän: `"and komoto.alkamisvuosi = ? "`, tai kenties tälle nullien karsimiselle on jokin suorituskykysyy?

Voisiko kyselyn tehdä unionin sijaan yhdellä kutsulla or-lausekkeella? Esim. näin:

```
(select distinct(h3.*)
 from koulutusmoduuli_toteutus komoto
   join koulutus_hakukohde kh on komoto.id = kh.koulutus_id
   join hakukohde h2 on kh.hakukohde_id = h2.id
   join haku h3 on h2.haku_id = h3.id
 where h2.tila = 'JULKAISTU'
       and komoto.tila = 'JULKAISTU'
       and ((h3.koulutuksen_alkamisvuosi is null
             and komoto.alkamisvuosi = 2018) or (
              h3.koulutuksen_alkamisvuosi = 2018
            ))
);
```


Suorituskyvyssä tällä ei kyllä näyttänyt QA:lla olevan merkittävää eroa, mutta ehkä vähentää toistoa ja parantaa luettavuutta.